Add script to copy shared libaries into application bundle

treese opened the pull request:
New Makefile target prepinstall to run a script that copies shared libraries from Homebrew into the application bundle. Libraries are recompiled/reinstalled in homebrew as needed to have the same MAC OS deployment target as the Aquamacs build. 

Note: I'm going to be offline for a few days on vacation, and I'll look at anything problematic with it when I get back.

treese committed:
Add script to copy shared libaries into application bundle

seh commented:
I'm still searching for a workaround for Aquamacs's inability to open TLS connections. Is there a subset of this patch I can use against the released Aquamacs 3.4 to repair it?

treese commented:
@seh if you install gnutls with homebrew and add 
```
(with-eval-after-load 'tls
  (push "/private/etc/ssl/cert.pem" gnutls-trustfiles))
```
to your Emacs config before it tries to touch the package system for MELPA, I think that will work just fine.

Would that be an option?

rpgoldman commented:
You can do the same process with MacPorts, too, if that's your flavor of package manager.

Sent from my iPhone

> On Sep 4, 2018, at 20:17, Win Treese <notifications@github.com> wrote:
> 
> @seh if you install gnutls with homebrew and add
> 
> (with-eval-after-load 'tls
>   (push "/private/etc/ssl/cert.pem" gnutls-trustfiles))
> to your Emacs config before it tries to touch the package system for MELPA, I think that will work just fine.
> 
> Would that be an option?
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or mute the thread.

seh commented:
@treese, that worked perfectly. Thank you!

Is it the case that with your patch, I wouldn't need to install the "gnutls" Homebrew package explicitly?

treese commented:
@seh That's right--the goal of the patch is to make this a non-issue for Aquamacs users, especially those who don't want to fiddle around with things like homebrew. One of the nice things about Aquamacs is that it reaches an audience that isn't as interested in messing around with their systems, or even much of the configuration.

seh commented:
That's good news.

Now, who can we get to review this patch? I took a look and it looks like well-crafted shell code, but I'm not familiar enough with the problem context to weigh in.

davidswelt commented:
I have two questions so far:  What does the script do when homebrew isn't installed?  Why is it called "prepinstall" when "post-install" or rather, "install-brewed-libraries" would be the more fitting name for the Makefile target?

treese committed:
Rename prepinstall target to install-brewed-libraries

treese committed:
Check for Homebrew installation before trying to copy libraries.

Added status messages about what is happening during build.

treese commented:
Good points. I updated the pull request to:
1. Use install-brewed-libraries as the Makefile target
2. Modified the script to check for Homebrew installation, and exit (successfully, for make) if not.

davidswelt committed:
fix typo

davidswelt commented:
OK, let's give this a shot for the upcoming 3.5 release.

davidswelt commented:
Setting Aquamacs version: 3.5
Find the Aquamacs build in nextstep/Aquamacs.app
If Homebrew is installed, build and copied brewed
shared libraries for distribution
sh aquamacs/build/build-homebrew-libraries.sh  \
	   nextstep/Aquamacs.app
mkdir: /Contents/MacOS: No such file or directory
make: *** [install-brewed-libraries] Error 1
janelle:aquamacs-emacs-3.4 dr$

davidswelt commented:
Looks like this is because $MACOSX_DEPLOYMENT_TARGET is not defined on my local build machine.  The code should handle that appropriately.

treese commented:
Hi, David. Sorry about that. It raises an interesting question about how you want it to behave.

The problem with the homebrew libraries arises when a distribution version is built that may be installed on systems without the same homebrew libraries.  In that situation, I assumed that the whole process was driven by the build scripts, which define MACOSX_DEPLOYMENT_TARGET.

That's not the right assumption for a development machine, which is what I missed. I think the right fix for that is to simply not bother fiddling with homebrew at all, because the libraries will obviously be present on the development machine.

I'm happy to make that change (or any other way you want it to work). Let me know.

davidswelt commented:
Hi Win,

“make install” should not globally install libraries with home-brew or in any other way.  It should install into nextstep/Aquamacs.app and not affect anything else.

I don’t think we need to include this process in “make install” to begin with.  If absolutely necessary, the post processing target that you wrote can be used, and I’ll call it on the official build machine.

Second, I got this error:

> * Process nextstep/Aquamacs.app/Contents/MacOS/lib/libgnutls.30.dylib
> rebuilding libunistring for minimum version /usr/local/opt/libunistring/lib/libunistring.2.dylib
> Error: libunistring: /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/libunistring.rb:41: syntax error, unexpected ',', expecting keyword_end
>                    "-o", "test"
>                         ^
> make: *** [install-brewed-libraries] Error 1


After a brew update, I called "make install-brewed-libraries”, but this time it didn’t even try to reinstall gnutls.  What gives?
Some debugging revealed that $(get_local_libs ${target}) is empty.   I think your awk call in get_local_libs returns nothing.


D

treese commented:
1. You're right -- no need to bother with this in the Makefile, as long as it is run on the build machine for distribution. Sorry for being overzealous there.

2. The problem with building libunistring was that I was a bit sloppy in modifying the Homebrew recipes. The next pull request fixes that.

3. When you did the brew update and ran it, had you rebuilt Aquamacs or not? If not, then the problem is likely that the script tried to be clever by not rebuilding a library if it was already in the build directory, but I didn't think about the situation of a brew update without an Aquamacs rebuild. In the new pull request, the script always processes the library. The pull request fixes that as well.

Thanks for your patience on this--it's a fragile approach, but I think it's the best one we have.

treese commented:
David: I was going through a fresh build with the latest aquamacs3 branch, and it is picking up the libraries:

jpeg
librsvg
glib
gdk-pixbuf
glib
cairo

from my homebrew in addition to gnutls. Is that something you expect to see now? Or just an artifact of what I have installed now?

davidswelt commented:
I’m not sure. Did we change the path protection in the build script? That would cause it.
S
On Oct 19, 2018, 22:59 -0400, Win Treese <notifications@github.com>, wrote:
> David: I was going through a fresh build with the latest aquamacs3 branch, and it is picking up the libraries:
> jpeg
> librsvg
> glib
> gdk-pixbuf
> glib
> cairo
> from my homebrew in addition to gnutls. Is that something you expect to see now? Or just an artifact of what I have installed now?
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub, or mute the thread.

treese commented:
If by "path protection" you mean the part in the build script that explicitly sets PATH to avoid MacPorts, etc. That's still there. But it doesn't protect against a library being picked up. Since I have both libjpeg and librsvg in my homebrew, that's probably why.

Which makes me think of a couple of other things:

1. If you want set CFLAGS=-mmacosx-version-min=${MIN_VERSION} (for the version you want), and set HOMEBREW_BUILD_FROM_SOURCE we could avoid tweaking the recipes individually, because the right versions would always be present, even after rebuilds. Then we just need to copy the right libraries. (It would make sense to document this somewhere for anyone in the future who is trying to build a distribution version.)

2. Because this mechanism is general for libraries, it makes it possible for Aquamacs to use other libraries that don't ship with MacOS but that Emacs can use (like librsvg), and you can consider adding some of those to the build in a way that couldn't be done before.


2.

davidswelt commented:
Here’s my suggestion.

Use your makefile target to build or update the libraries as needed using home-brew, then copy them (I.e., binaries) into the checkout in a certain place, such as aquamacs/lib.
We will make these binaries part of the source distribution.

Adjust the build process such as this aquamacs/lib folder is included and can be pulled from.

It’s always okay to include binaries, but obviously we need to avoid is linking to installed libraries in the build system (that may be missing from the user’s runtime environment).


Also, what you’re telling me is to uninstall (with brew) some libraries from the build machine, right?

David





> On Oct 20, 2018, at 5:10 PM, Win Treese <notifications@github.com> wrote:
> 
> If by "path protection" you mean the part in the build script that explicitly sets PATH to avoid MacPorts, etc. That's still there. But it doesn't protect against a library being picked up. Since I have both libjpeg and librsvg in my homebrew, that's probably why.
> 
> Which makes me think of a couple of other things:
> 
> 	• If you want set CFLAGS=-mmacosx-version-min=${MIN_VERSION} (for the version you want), and set HOMEBREW_BUILD_FROM_SOURCE we could avoid tweaking the recipes individually, because the right versions would always be present, even after rebuilds. Then we just need to copy the right libraries. (It would make sense to document this somewhere for anyone in the future who is trying to build a distribution version.)
> 
> 	• Because this mechanism is general for libraries, it makes it possible for Aquamacs to use other libraries that don't ship with MacOS but that Emacs can use (like librsvg), and you can consider adding some of those to the build in a way that couldn't be done before.
> 
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

treese commented:
I've got a new version that separates:
1. Copying necessary shared libs into the Aquamacs bundle for distribution
2. Rebuilding homebrew libraries to match the minimum version needed

I think this satisfies the (good!) requirement that the build process not mess with other things on the system. It also means that you don't have to constantly rebuild things on your build-distribution machine, since we can check the min version in those. Almost no one else needs to worry about those version checks.

We might want to tweak build-aquamacs so the default build does not enforce a minimum version, and that you force it on the build-distribution system.

It occurred to me that your goal might be to let someone build Aquamacs without homebrew at all. To do that, we'd have to start copying header files, etc. along with library binaries. That seems hard.

Finally, on things like libsrvg: configure picks them up automatically from homebrew if they happen to be installed. I tried a test configuration by changing build-aquamacs to call `configure` with `--without-rsvg`, which turned it off. So it looks like you need to be careful with what's installed on the build machine, or with configure options. On the other hand, if you wanted to enable (say) librsvg, the mechanisms we have for gnutls will work just fine for that as well, so you have new configuration options possible.

davidswelt commented:
Is that revision available somewhere?
Yes, one would want to be able to build Aquamacs with and without homebrew.  If libraries are available, they are to be copied into place, and if not, an extra build target that can be called from build.sh would show warnings.  
I agree - the build should not enforce minimum versions.  Perhaps if they aren't compatible, don't link, show a warning.  Provide a separate script or build target that will build the homebrew libraries and even install them.

davidswelt commented:
Ping -- pointer to revision?
